src/zns/utils.ts

```diff
@@ -0,0 +1,264 @@
+const GENERATOR = [0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3];
+// HRP is the human-readable part of zilliqa bech32 addresses
+const HRP = 'zil';
+const tHRP = 'tzil';
+
+function isByteString (str: string, len: number) {
+  return !!str.replace('0x', '').match(`^[0-9a-fA-F]{${len}}$`);
+};
+
+function isAddress (address: string) {
+  return isByteString(address, 40);
+};
+
+/**
+ * convertBits
+ *
+ * groups buffers of a certain width to buffers of the desired width.
+ *
+ * For example, converts byte buffers to buffers of maximum 5 bit numbers,
+ * padding those numbers as necessary. Necessary for encoding Ethereum-style
+ * addresses as bech32 ones.
+ *
+ * @param {Buffer} data
+ * @param {number} fromWidth
+ * @param {number} toWidth
+ * @param {boolean} pad
+ * @returns {Buffer|null}
+ */
+function convertBits (
+  data: Buffer,
+  fromWidth: number,
+  toWidth: number,
+  pad: boolean = true,
+) {
+  let acc = 0;
+  let bits = 0;
+  const ret = [];
+  const maxv = (1 << toWidth) - 1;
+  // tslint:disable-next-line
+  for (let p = 0; p < data.length; ++p) {
+    const value = data[p];
+    if (value < 0 || value >> fromWidth !== 0) {
+      return null;
+    }
+    acc = (acc << fromWidth) | value;
+    bits += fromWidth;
+    while (bits >= toWidth) {
+      bits -= toWidth;
+      ret.push((acc >> bits) & maxv);
+    }
+  }
+
+  if (pad) {
+    if (bits > 0) {
+      ret.push((acc << (toWidth - bits)) & maxv);
+    }
+  } else if (bits >= fromWidth || (acc << (toWidth - bits)) & maxv) {
+    return null;
+  }
+
+  return Buffer.from(ret);
+};
+
+function hrpExpand (hrp: string): Buffer {
+  const ret = [];
+  let p;
+  for (p = 0; p < hrp.length; ++p) {
+    ret.push(hrp.charCodeAt(p) >> 5);
+  }
+  ret.push(0);
+  for (p = 0; p < hrp.length; ++p) {
+    ret.push(hrp.charCodeAt(p) & 31);
+  }
+  return Buffer.from(ret);
+};
+
+function polymod (values: Buffer): number {
+  let chk = 1;
+  // tslint:disable-next-line
+  for (let p = 0; p < values.length; ++p) {
+    const top = chk >> 25;
+    chk = ((chk & 0x1ffffff) << 5) ^ values[p];
+    for (let i = 0; i < 5; ++i) {
+      if ((top >> i) & 1) {
+        chk ^= GENERATOR[i];
+      }
+    }
+  }
+  return chk;
+};
+
+function createChecksum(hrp: string, data: Buffer) {
+  const values = Buffer.concat([
+    Buffer.from(hrpExpand(hrp)),
+    data,
+    Buffer.from([0, 0, 0, 0, 0, 0]),
+  ]);
+  // var values = hrpExpand(hrp).concat(data).concat([0, 0, 0, 0, 0, 0]);
+  const mod = polymod(values) ^ 1;
+  const ret = [];
+  for (let p = 0; p < 6; ++p) {
+    ret.push((mod >> (5 * (5 - p))) & 31);
+  }
+  return Buffer.from(ret);
+}
+
+
+function verifyChecksum(hrp: string, data: Buffer) {
+  return polymod(Buffer.concat([hrpExpand(hrp), data])) === 1;
+}
+
+function encode(hrp: string, data: Buffer) {
+  const combined = Buffer.concat([data, createChecksum(hrp, data)]);
+  let ret = hrp + '1';
+  // tslint:disable-next-line
+  for (let p = 0; p < combined.length; ++p) {
```

Comment:
Should we use `for ... of` maybe? Is it that tslint suggests?
(the same about all the other c-style for loops)

Reply:
We copied this code from other library and we want it be in its original state.